presto-main/src/main/java/com/facebook/presto/execution/executor/MultilevelSplitQueue.java

```diff
@@ -73,23 +74,39 @@ public MultilevelSplitQueue(boolean levelAbsolutePriority, double levelTimeMulti
 
     private void addLevelTime(int level, long nanos)
     {
-        lock.lock();
-        try {
-            levelScheduledTime[level] += nanos;
-        }
-        finally {
-            lock.unlock();
-        }
+        levelScheduledTime[level].addAndGet(nanos);
     }
 
+    /**
+     * During periods of time when a level has no waiting splits, it will not accumulate
+     * scheduled time and will fall behind relative to other levels.
+     * <p>
+     * This can cause temporary starvation for other levels when splits do reach the
+     * previously-empty level.
+     * <p>
+     * To prevent this we set the scheduled time for levels which were empty to the expected
+     * scheduled time.
+     */
     public void offer(PrioritizedSplitRunner split)
     {
         checkArgument(split != null, "split is null");
 
         split.setReady();
+        int level = split.getPriority().getLevel();
         lock.lock();
         try {
-            levelWaitingSplits.get(split.getPriority().getLevel()).offer(split);
+            if (levelWaitingSplits.get(level).isEmpty()) {
+                // Accesses to levelScheduledTime are not synchronized, so we have a data race
+                // here - our level time math will be off. However, the staleness is bounded by
+                // the fact that only running splits that complete during this computation
+                // can update the level time. Therefore, this is benign.
+                long level0Time = getLevel0TargetTime();
+                long levelExpectedTime = (long) (level0Time / Math.pow(levelTimeMultiplier, level));
+                long delta = levelExpectedTime - levelScheduledTime[level].get();
```

Comment:
Because that seems like it would lead to a worse data race - we would permanently be behind where we should be.